### ISSUE TITLE
Don't use MediaInfo "Strings"

### ISSUE THREAD

**Issue by Sami32** [DigitalMediaServer/DigitalMediaServer#17]
Not using `/String` MediaInfo parameters will enhance the parsing speed.
And some more optimizations are also possible, that give also some additional ADTS bitrate informations.

From my test, we can win around 5% speed improvement.

**Comment by Nadahar** [comment]
@Sami32 I don't understand what you mean, I think you need to explain a bit better.

**Comment by Sami32** [comment]
Example:
```
ID                        : The ID for this stream in this file
ID/String                 : The ID for this stream in this file
OriginalSourceMedium_ID   : The ID for this stream in the original medium of the material
OriginalSourceMedium_ID/S : The ID for this stream in the original medium of the material
UniqueID                  : The unique ID for this stream, should be copied with stream copy
UniqueID/String           : The unique ID for this stream, should be copied with stream copy
MenuID                    : The menu ID for this stream in this file
MenuID/String             : The menu ID for this stream in this file
```

Do not use the `/String` ones, as `ReadByHuman` could be used and make the parsing around 5% faster.

That's said 3 others MediaInfo optimizations are also possible that still enhance a little the speed and give ADTS bitrate infos without speed losses. 

I'm afraid that i cannot be more clear that that.

**Comment by Nadahar** [comment]
@Sami32 That helps, although I don't know the difference between e.g ```ID``` and ```ID/String```. What data i returned in ```ID```? Obviously if we can avoid using strings that is not only faster but better (type safety etc,). Does MediaInfo offer methods that return integer/long types?

**Comment by Sami32** [comment]
The `/String` ones are human readable. Example:
`ScanOrder` TFF
`ScanOrder/String` Top Field First"

Yes, i know that you like verbose, but when it is not necessary and that speed parsing is "lost"...

**Comment by Nadahar** [comment]
@Sami32 I don't think the parsing speed really matters, it's very much faster than many other things that's being done anyway. When it comes to parsing, I prefer precision, not "verbose" strings. Most of all, I prefer anything that isn't a string to not go via string, as it's an unnecessary transformation with a relatively high risk of bugs.

**Comment by Sami32** [comment]
Well, when i read the users complaining about parsing time i guess that they don't share, as me, your view.

Whatever, if i understood you correctly i can leave this "issue" open ?

**Comment by Sami32** [comment]
Well, my memory give me trouble, so i did verify and find this special case to answer you:
```
ID : 49-CC1
ID : 49 (0x31)-CC1
```
https://github.com/UniversalMediaServer/UniversalMediaServer/issues/1232

**Comment by Nadahar** [comment]
@Sami32 Sure, it sounds like something that would be an improvement so it should stay. I'm not quite sure of the details yet though.

**Comment by Sami32** [comment]
I did some searches and test and commit sometime ago, but as i was feeling that i was the only one considering it, i didn't continued on that.
You could find some trace of it [there](https://github.com/UniversalMediaServer/UniversalMediaServer/compare/MediaInfo_test).

**Comment by Nadahar** [comment]
@Sami32 "Parsing" is an imprecise term that can mean many things. I suspect that when users complain about parsing time, it's the whole process - most of which is I/O bound when MediaInfo is reading the media file. I'm not saying that this "parsing time" doesn't matter, I'm saying that whatever you save on parsing a shorter string is in the sub millisecond range and contributes a negligible part of the time spent "parsing" a given media item.

**Comment by Sami32** [comment]
Well, some users have between 20 et 50 thousands media files, so i guess that make some difference to them ? From what i read, the complain came mostly from users havings at least 13 or 19 thousand media files.

**Comment by Nadahar** [comment]
@Sami32 I can't imagine the difference is anywhere near 1 millisecond, but if it were, 50 thousand milliseconds, it would constitue 50 seconds. The time to scan 50 000 media files on the other hand, will take hours, so I really doubt that 50 seconds would make much of a difference. In reality it will be much less though, so the total delay is probably closer to 1 second for 50 000 media files.

I think there are lots of other things that could be done that would be much more effective, like preventing media from being scanned many times.

**Comment by Sami32** [comment]
OK, i see, i didn't know that was that time scale. Thanks.
I only know that when i didn't used the `Sting`, i got 5% parsing speed increase, after repeating many tests, so i was only referring to my real tests; so that doesn't seem coherent with what you are saying.
Perhaps as these values was not used, the multiple parsing was avoid for all the `/String` and that's what make a such difference ?

Yes, i agree, as you've already said, the multiple time scanning, due to chaotic organisation, no enum class, is a pain to maintain, enhance...

**Comment by Nadahar** [comment]
@Sami32 It depends a bit on how MediaInfo functions internally - it might be that simply querying the "/String" values is slower than querying the others (without us doing any parsing). I don't know if MediaInfo sets all the values when parsing the media file and then simply retrieves them when queried, or if the information is processed before retrieval. The only way to really know is to study the MediaInfo code or do tests on simply retrieving values.

In any case, I think the most important thing when parsing is to choose the most "precise and predictable" input value - where the chance of parsing errors are least. Human readable text fields isn't my first choice for parsing, so rewriting it to use the non "/String" might be worth it for improved "quality" of data alone.

It doesn't have to be one or the other either, we could change some and leave some be, depending on the nature of the data and the expected values.

**Comment by Sami32** [comment]
Reading you make me  realize that MediaInfo is parsing all the values, and when i set it to false for the human readable ones and some others as well, it just read much less values.
Now i remember the comments on his owner in Sourceforge.
So the 5% speed increase came only from MediaInfo parsing only the minimum values requested.
Thanks for your insight ;-)

Well, yes, but only avoiding them totally will make that MediaInfo setting possible and enhance the speed. But as you said, perhaps that a better Java MediaInfo values parsing handling could even make the speed enhancement better than 5% by itself ? I really don't know.

**Comment by Nadahar** [comment]
@Sami32 I don't think there's much speed to be gained by the changing the Java handling, although we could certainly gain some nanoseconds here and there. The more important thing as I see it, is still to get the most precise parsing (where you know exactly what to expect). If disabling this can both make MediaInfo run 5% faster *and* give more precise parsing, it shouldn't be a very difficult choice though.

**Comment by Sami32** [comment]
Yes, i read that since the beginning you are seing from the parsing precision side and me the parsing speed side :wink: 
Two in one seem a good "compromise" :rofl:

**Comment by Nadahar** [comment]
I think this should be a part of a larger parsing refactoring.
